doc/api/async_hooks.md

```diff
@@ -0,0 +1,532 @@
+
+* Returns {number} the `id` of the current execution context. Useful to track
+  when something calls.
+
+For example:
+
+```js
+console.log(async_hooks.currentId());  // 1 - bootstrap
+fs.open(path, (err, fd) => {
+  console.log(async_hooks.currentId());  // 2 - open()
+});
+```
+
+It is important to note that the ID returned fom `currentId()` is related to
+execution timing, not causality (which is covered by `triggerId()`). For
+example:
+
+```js
+const server = net.createServer(function onConnection(conn) {
+  // Returns the ID of the server, not of the new connection, because the
+  // onConnection callback runs in the execution scope of the server's
+  // MakeCallback().
+  async_hooks.currentId();
+
+}).listen(port, function onListening() {
+  // Returns the ID of a TickObject (i.e. process.nextTick()) because all
+  // callbacks passed to .listen() are wrapped in a nextTick().
+  async_hooks.currentId();
+});
+```
+
+#### `async_hooks.triggerId()`
+
+* Returns {number} the ID of the resource responsible for calling the callback
+  that is currently being executed.
+
+For example:
+
+```js
+const server = net.createServer((conn) => {
+  // Though the resource that caused (or triggered) this callback to
+  // be called was that of the new connection. Thus the return value
+  // of triggerId() is the ID of "conn".
+  async_hooks.triggerId();
+
+}).listen(port, () => {
+  // Even though all callbacks passed to .listen() are wrapped in a nextTick()
+  // the callback itself exists because the call to the server's .listen()
+  // was made. So the return value would be the ID of the server.
+  async_hooks.triggerId();
+});
+```
+
+## Embedder API
+
+Library developers that handle their own I/O, a connection pool, or
+callback queues will need to hook into the AsyncWrap API so that all the
+appropriate callbacks are called. To accommodate this a JavaScript API is
+provided.
+
+### `class AsyncResource()`
+
+The class `AsyncResource` was designed to be extended by the embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
+
+The `init()` hook will trigger when an `AsyncResource` is instantiated.
+
+It is important that before/after calls are unwined
+in the same order they are called. Otherwise an unrecoverable exception
+will occur and node will abort.
+
+```js
+// AsyncResource() is meant to be extended. Instantiating a
+// new AsyncResource() also triggers init(). If triggerId is omitted then
+// async_hook.currentId() is used.
+const asyncResource = new AsyncResource(type, triggerId);
+
+// Call AsyncHooks before callbacks.
+asyncResource.emitBefore();
+
+// Call AsyncHooks after callbacks.
+asyncResource.emitAfter();
+
+// Call AsyncHooks destroy callbacks.
+asyncResource.emitDestroy();
+
+// Return the unique ID assigned to the AsyncResource instance.
+asyncResource.asyncId();
+
+// Return the trigger ID for the AsyncResource instance.
+asyncResource.triggerId();
+```
```

Comment:
If this is supposed to be an API overview then it should probably have the sub-heading that says that. But it may also want to be removed, like I discussed above or the other similar section.